Python/marshal.c

```diff
@@ -503,9 +503,41 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
             W_TYPE(TYPE_SET, p);
         n = PySet_GET_SIZE(v);
         W_SIZE(n, p);
+        // bpo-37596: We need to write the elements in a deterministic order!
+        // Since we know that they all need marshal support anyways, this can
+        // be conveniently accomplished by using their marshal serializations
+        // as sort keys:
+        PyObject *pairs = PyList_New(0);
+        if (pairs == NULL) {
+            p->error = WFERR_NOMEMORY;
+            return;
+        }
         while (_PySet_NextEntry(v, &pos, &value, &hash)) {
+            PyObject *dump = PyMarshal_WriteObjectToString(value, p->version);
+            if (dump == NULL) {
+                p->error = WFERR_UNMARSHALLABLE;
+                goto anyset_done;
+            }
+            PyObject *pair = PyTuple_Pack(2, dump, value);
+            Py_DECREF(dump);
+            int error = pair == NULL || PyList_Append(pairs, pair);
+            Py_XDECREF(pair);
+            if (error) {
+                p->error = WFERR_NOMEMORY;
+                goto anyset_done;
+            }
+        }
+        if (PyList_Sort(pairs)) {
+            p->error = WFERR_UNMARSHALLABLE;
```

Comment:
What could trigger this error?  I'm thinking that the only failure possible at this point would be *WFERR_NOMEMORY*